sdks/python/apache_beam/runners/dataflow/internal/apiclient.py

```diff
@@ -301,6 +351,22 @@ def __init__(self, packages, options, environment_version, pipeline_url):
           dataflow.Environment.SdkPipelineOptionsValue.AdditionalProperty(
               key='display_data', value=to_json_value(items)))
 
+  def _get_environments_from_tranforms(self):
+    if not self._proto_pipeline:
+      return []
+    environment_ids = []
+    for transform in self._proto_pipeline.components.transforms.values():
+      if transform.environment_id not in environment_ids:
+        environment_ids.append(transform.environment_id)
+    environments = []
+    for environment_id in environment_ids:
+      if not environment_id:
```

Comment:
Nit: I'd move this up into the previous loop. Or I'd write this as a list comprehension, e.g. 

```
environment_ids = set(
    transform.environment_id
    for transform in self._proto_pipeline.components.transforms.values()
    if transform.environment_id)
return [self._proto_pipeline.components.environments[id] for id in environment_ids]
```